stan/math/rev/core/precomputed_gradients.hpp

```diff
@@ -17,51 +19,119 @@ namespace math {
  *
  * Stan users should use function precomputed_gradients()
  * directly.
+ *
+ * @tparam ContainerOperands tuple of any container operands (var_value
+ * containing Eigen types)
+ * @tparam ContainerGradients tupleof any container gradients (Eigen types)
  */
-class precomputed_gradients_vari : public vari {
+template <typename ContainerOperands = std::tuple<>,
+          typename ContainerGradients = std::tuple<>>
+class precomputed_gradients_vari_template : public vari {
  protected:
   const size_t size_;
   vari** varis_;
   double* gradients_;
+  static_assert(std::tuple_size<ContainerOperands>::value
+                    == std::tuple_size<ContainerGradients>::value,
+                "precomputed_gradients_vari: ContainerOperands and "
+                "ContainerGradients should have same size!");
+  static constexpr size_t N_containers
+      = std::tuple_size<ContainerOperands>::value;
+  ContainerOperands container_operands_;
+  ContainerGradients container_gradients_;
+
+  /**
+   * Checks that sizes of containers in container operands and
+   * container_gradients match.
+   *
+   * @throws std::invalid_argument sizes do not match
+   */
+  template <size_t... Is>
+  void check_sizes(std::index_sequence<Is...>) {
+    static_cast<void>(std::initializer_list<int>{
+        (check_size_match("precomputed_gradients_vari", "rows of operands",
+                          std::get<Is>(container_operands_).vi_->rows(),
+                          "rows of gradients",
+                          std::get<Is>(container_gradients_).rows()),
+         check_size_match("precomputed_gradients_vari", "cols of operands",
+                          std::get<Is>(container_operands_).vi_->cols(),
+                          "cols of gradients",
+                          std::get<Is>(container_gradients_).cols()),
+         0)...});
+  }
```

Comment:
Its a standard way of doing things in C++. I don't think we should avoid that, just because you are unfamilliar with it. That said, I don't mind your `for_each` approach, so we can replace this, once `for_each` is in.